pkgs/build-support/trivial-builders/default.nix

```diff
@@ -625,18 +625,24 @@ rec {
 
   # Docs in doc/build-helpers/trivial-build-helpers.chapter.md
   # See https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-writeReferencesToFile
-  writeReferencesToFile = path: runCommand "runtime-deps"
+  # TODO: Convert to throw after Nixpkgs 24.05 branch-off.
+  writeReferencesToFile = (if config.allowAliases then lib.warn else throw)
+    "writeReferencesToFile is deprecated in favour of writeReferenceClosureToFile"
+    (path: writeReferenceClosureToFile [ path ]);
+
+  # Docs in doc/build-helpers/trivial-build-helpers.chapter.md
+  # See https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-writeReferenceClosureToFile
+  writeReferenceClosureToFile = paths: runCommand "runtime-deps"
     {
-      exportReferencesGraph = [ "graph" path ];
+      __structuredAttrs = true; # Get the cleaner exportReferencesGraph interface
+      exportReferencesGraph.graph =
+        lib.sort lib.lessThan (
+          map (p: "${p}") (
+            lib.filter (p: p != null && p != "") paths));
```

Comment:
1. I imagine this means that in order to _evaluate_ anything about `exportReferencesGraph.graph` (e.g. `(head exportReferencesGraph.graph).pname)`, or length, or whatever) we first have to run the _builds_ for all of the `paths`. Idk if that fundamentally changes anything.
2. Can be rewritten as a flatter expression using e.g. `lib.pipe`

<details>

```
lib.pipe paths [
  (builtins.filter (p: p != null && p != "")
  (map (p: "${p}")) # or builtins.toString?
  (lib.sort lib.lessThan)
]
```

</details>

Reply:
> 1. I imagine this means that in order to _evaluate_ anything about `exportReferencesGraph.graph` (e.g. `(head exportReferencesGraph.graph).pname)`, or length, or whatever) we first have to run the _builds_ for all of the `paths`. Idk if that fundamentally changes anything.

These string-like stuff will get interpolated during derivation evaluation, even if I don't interpolate them manually. They are the *dependencies* of the result derivation, and will be realized before building the result derivation. That's essentially how Nix works.

> 2. Can be rewritten as a flatter expression using e.g. `lib.pipe`
> 
> <details>
> 
> ```
> lib.pipe paths [
>   (builtins.filter (p: p != null && p != "")
>   (map (p: "${p}")) # or builtins.toString?
>   (lib.sort lib.lessThan)
> ]
> ```
> 
> </details>

We cannot use `toString` instead of string interpolation in this context. `toString ./.` simply evaluates to the absolute path to that source directory instead of a store path. This case is covered under `tests.trivial-builders.references`, as one of the sample attribute `tests.trivial-builders.references.samples.path` is a Nix-Language path. I made similar mistake when working on #204462; it took me hours to debug.

As for `lib.pipe`, I like the style pretty much, but I'm not sure if other people would think of it as an unnecessary evaluation overhead or not.

Reply:
Alright, maintainability matters. Let's embrace a flatter stream-processing expression.